(Fix) [Contract Interaction] Parameter validation

fernandomg opened the pull request:
This PR closes #950, by:

- utilizing new `modifiedSinceLastSubmit` final-form's flag
- identifying failing values and marking the corresponding field
- displaying generic errors when we can not identify a field in particular
  - this required a new component to display errors in the form
  - to proactively identify a failing tx, it's calling `estimateGas`, thus the generic error message
- hiding 'Value' field, when methods are not payable

This solves **Report 2** and **Report 3** listed in the issue. I'll be working on a separate PR to solve **Report 1** (#975).

~**note:** found a bug, if you make it to the 'review screen', then go back, select a different method and review it again, data displayed (and submitted) is the one that was generated for the previous review.~

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

fernandomg commented:
> note: found a bug, if you make it to the 'review screen', then go back, select a different method and review it again, data displayed (and submitted) is the one that was generated for the previous review.

This was a simple fix so far [420b9cb](https://github.com/gnosis/safe-react/pull/974/commits/420b9cb58e33564be57dc40583b84ef0a3453892)

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

mikheevm commented on src/routes/safe/components/Balances/SendModal/screens/ContractInteraction/Buttons/index.tsx:
```diff
@@ -5,70 +5,41 @@ import { useField, useFormState } from 'react-final-form'
-import { createTxObject } from 'src/routes/safe/components/Balances/SendModal/screens/ContractInteraction/utils'
+import { isReadMethod } from 'src/routes/safe/components/Balances/SendModal/screens/ContractInteraction/utils'
 
 const useStyles = makeStyles(styles as any)
```
is it possible to get rid of `any` there?

fernandomg replied:
Ok, this was fun [f049f85](https://github.com/gnosis/safe-react/pull/974/commits/f049f8598d5b6f3c5d9d9d3d97aad54f9ca3986d)

https://material-ui.com/styles/api/#createstyles-styles-styles

mikheevm commented on src/routes/safe/components/Balances/SendModal/screens/ContractInteraction/Buttons/index.tsx:
```diff
@@ -5,57 +5,41 @@ import { useField, useFormState } from 'react-final-form'
 const useStyles = makeStyles(styles as any)
 
-const Buttons = ({ onCallSubmit, onClose }) => {
+const Buttons = ({ onClose }) => {
```
could you please add prop types for this component?

fernandomg replied:
done, also added props to most of the Contract Interaction components

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

Agupane commented on src/logic/contractInteraction/sources/ABIService/index.ts:
```diff
@@ -0,0 +1,38 @@
+  return `${name}(${params})`
+}
+
+export const getSignatureHash = (signature) => {
```
Could you please add the types for all this getter methods?

fernandomg commented:
[e3a9945](https://github.com/gnosis/safe-react/pull/974/commits/e3a9945675e6241a4b48a88154d1ba2a77c56f84):
- Found several bugs (well, missuses for `useField` hook)
- Also removed the `types.d.ts` for `ABIService` and used the `AbiItem` type from `web3-utils`

[0cf786e](https://github.com/gnosis/safe-react/pull/974/commits/0cf786e8d1ba2aa03b40a44b71ab02d66fb5c424):
- Found and fixed a bug related to the `data` string not being updated in the review screen after modifying it
<br>
Sorry @mikheevm, @Agupane. Can you re-review this?

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

fernandomg commented:
> The ENS forward resolution does not work, but this also wasn't implemented in before, right? If this is easy to implement it I would suggest adding it in this PR or a separate one. If it takes more time I would change the error message in this PR and create a ticket to prioritize it for later.

Indeed, it never was implemented. I'll create a new issue to add that, so we better focus on the validation feature.

It should be fairly simple to do.

francovenica commented:
In order to check this I had to switch to this branch and point my local env to prod so I could test the same case the user reported, with the same contract and method

The fields now are showing the error messages, the specific case with the byte32 field now works, and the review button is not getting stuck into a "grayed out" state

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app

fernandomg commented:
trying to restart travis build, PR hangs in _pending_ status: https://github.com/travis-ci/travis-ci/issues/9789

gnosis-info commented:
Travis automatic deployment:
 https://pr974--safereact.review.gnosisdev.com/app
